### PR TITLE
Update lighting setup for higher quality rendering

### PR DESCRIPTION
**Pull request by ericjansenrpm**
Ticket: [Visage Rendering Quality](https://ready-player-me.atlassian.net/browse/ACT-342?atlOrigin=eyJpIjoiYzVkZTYxYTJkOWI0NDhlOTgzNjg4ZGY5MDA5ZTEyYzkiLCJwIjoiaiJ9)

- update lighting setup to match the render API
- fix ACES tone mapping
- switch HDRI to soft studio
- align docs in storybook
- expose new lighting props in avatar component
- add optional vignette effect that can be toggled

**Please note that the updated LightingProps type introduce a breaking change as discussed with Elar.**

Screenshot from 2023-12-18 13-31-29

We did some exploration with the background but for now stayed with the same setup since there is an option for the developer in studio to switch between light, dark and custom colors.

### CODE REVIEW

File: src/components/Lights/Lights.component.tsx
Comment (ericjansenrpm) [inline src/components/Lights/Lights.component.tsx:1]: Maybe start here since this components holds now all the lights. :)

File: src/services/Models.service.tsx [hunk 1]
```diff
@@ -79,7 +79,6 @@ export const normaliseMaterialsConfig = (materials: Record<string, Material>, bl
     if (mat.map) {
       mat.map.minFilter = LinearFilter;
       mat.depthWrite = true;
-      mat.toneMapped = false;
```
Comment (ericjansenrpm) [inline src/services/Models.service.tsx:82]: This fixed the tonemapping for materials / the avatar.

File: src/services/Environment.service.ts [hunk 2]
```diff
@@ -9,7 +9,8 @@ export const environmentPresets = {
   studio: 'studio',
   city: 'city',
   park: 'park',
-  lobby: 'lobby'
+  lobby: 'lobby',
+  soft: 'soft'
```
Comment (ericjansenrpm) [inline src/services/Environment.service.ts:13]: This is the new HDRI wich lays in the AWS bucket at around 1.9mb.

File: src/types/index.ts [hunk 3]
```diff
@@ -146,6 +155,10 @@ export type EffectConfiguration = {
    * Bloom post-processing effect.
    */
   bloom?: BloomConfiguration;
+  /**
+   * Vignette post-processing effect.
+   */
+  vignette?: boolean;
```
Comment (ericjansenrpm) [inline src/types/index.ts:161]: Optional vignette toggle.

File: src/components/Scene/CameraLighting.component.tsx
Comment (ericjansenrpm) [inline src/components/Scene/CameraLighting.component.tsx:1]: Moved lights out of the camera component to create a better seperation. In the future we might want to rename this component.

File: src/components/BaseCanvas/BaseCanvas.component.tsx [hunk 4]
```diff
@@ -27,7 +27,7 @@ export const BaseCanvas: FC<BaseCanvasProps> = ({
     key={fov}
     className={`${styles['base-canvas']} ${className ?? ''}`}
     shadows="soft"
-    gl={{ preserveDrawingBuffer: true, toneMappingExposure: 0.5, alpha: true }}
+    gl={{ preserveDrawingBuffer: true, alpha: true, toneMappingExposure: 1.4, toneMapping: ACESFilmicToneMapping }}
```
Comment (ericjansenrpm) [inline src/components/BaseCanvas/BaseCanvas.component.tsx:30]: Adjusted tonemapping for best contrast and skintone details.

File: src/services/Environment.service.ts (hunk 2 above)
Comment (blanxii) [reply src/services/Environment.service.ts:13]: can we optimise maybe? as we did in the past? 1.9mb is quite a lot..

File: src/types/index.ts [hunk 5]
```diff
@@ -20,16 +20,25 @@ export interface EnvironmentProps {
...
-  ambientLightIntensity?: number;
-  dirLightPosition?: Vector3;
-  dirLightColor?: Color | string;
-  dirLightIntensity?: number;
-  spotLightPosition?: Vector3;
-  spotLightColor?: string;
-  spotLightAngle?: number;
-  spotLightIntensity?: number;
+  keyLightIntensity?: number;
+  keyLightColor?: string; // css-color string
```
Comment (blanxii) [inline src/types/index.ts:28]: css-color string? is the same as hexadecimal color?

File: src/components/Avatar/Debug.stories.tsx [hunk 6]
```diff
@@ -33,15 +33,15 @@ Debug.args = {
...
-  ambientLightColor: '#fff5b6',
-  dirLightColor: '#002aff',
-  spotLightColor: '#fff5b6',
-  ambientLightIntensity: 0.25,
-  dirLightIntensity: 5,
-  spotLightIntensity: 1,
-  dirLightPosition: new Vector3(-3, 5, -5),
-  spotLightPosition: new Vector3(12, 10, 7.5),
-  spotLightAngle: 0.314,
+  keyLightIntensity: 1.2,
```
Comment (blanxii) [inline src/components/Avatar/Debug.stories.tsx:36]: dos it make sense to expose the LIGHT_CONFIG and just use this values? I can imagine doing some tuning on that config that we forget to update on the storybook, what do you think?

File: src/types/index.ts (hunk 5 above)
Comment (ericjansenrpm) [reply src/types/index.ts:28]: I think storybook provides this in #000000 where the three js color prefers 0x00000 but both should work :)

File: src/services/Environment.service.ts (hunk 2 above)
Comment (ericjansenrpm) [reply src/services/Environment.service.ts:13]: If we cut the resolution in half down to 512x256 we can get to 530kb

File: src/components/Scene/CameraLighting.component.tsx [hunk 7]
```diff
@@ -1,11 +1,10 @@
 import { useEffect, FC, useRef } from 'react';
 import { useFrame, useThree } from '@react-three/fiber';
-import { Camera, Vector3, DirectionalLight, AmbientLight, SpotLight } from 'three';
+import { Camera, Vector3 } from 'three';
 import { OrbitControls } from 'three-stdlib';
 import { clamp, lerp } from 'src/services';
-import { LightingProps, Required } from 'src/types';
 
-type CameraLightingProps = Required<LightingProps> & {
+type CameraLightingProps = {
```
Comment (Zaehiel) [inline src/components/Scene/CameraLighting.component.tsx:7]: This component is no longer handling lighting effects.
It would be good to rename it so we don't create confusion for the future.
e.g. `CameraControls`

File: src/services/Models.service.tsx [hunk 8]
```diff
@@ -79,7 +79,10 @@ export const normaliseMaterialsConfig = (materials: Record<string, Material>, bl
...
+    }
+
+    if (mat.name.toLowerCase().includes('hair')) {
+      mat.roughness = 0.9;
```
Comment (Zaehiel) [inline src/services/Models.service.tsx:85]: I'm wondering how come we have to modify roughness for hair explicitly?

File: src/components/Lights/Lights.component.tsx [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+  } as Required<LightingProps>
+});
+
+Object.fromEntries = (arr: Array<any>) => Object.assign({}, ...arr.map(([k, v]) => ({ [k]: v })));
+
+const definedProps = (obj: Object) =>
+  Object.fromEntries(
+    // eslint-disable-next-line
+    Object.entries(obj).filter(([k, v]) => v !== undefined)
+  );
```
Comment (Zaehiel) [inline src/components/Lights/Lights.component.tsx:32]: Why is this necessary?

File: src/components/Lights/Lights.component.tsx [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    keyLightIntensity,
+    keyLightColor,
+    fillLightIntensity,
+    fillLightColor,
+    fillLightPosition,
+    backLightIntensity,
+    backLightColor,
+    backLightPosition,
+    lightTarget
+  } = Object.assign(LIGHT_CONFIG.defaultProps, definedProps(lightingProps));
```
Comment (Zaehiel) [inline src/components/Lights/Lights.component.tsx:46]: Maybe we don't need the extra methods?
I was thinking it could be done like so

```typescript
const resolvedProps = {
...LIGHT_CONFIG.defaultProps,
...lightingProps
}

const {
    keyLightIntensity,
    keyLightColor,
    fillLightIntensity,
    fillLightColor,
    fillLightPosition,
    backLightIntensity,
    backLightColor,
    backLightPosition,
    lightTarget
  } = resolvedProps;

File: src/components/Lights/Lights.component.tsx [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+      />
+      {/* Lift light that creates soft light on body and shoes. */}
+      <spotLight
+        position={LIGHT_CONFIG.liftLightPosition}
+        target={targets.shoe}
+        angle={LIGHT_CONFIG.keyLightAngle}
+        color={keyLightColor}
+        intensity={keyLightIntensity * 0.16}
+      />
+      <ContactShadows opacity={1.0} scale={2.0} blur={1.0} far={0.25} resolution={256} color="#000000" />
```
Comment (Zaehiel) [inline src/components/Lights/Lights.component.tsx:99]: We should remove the other shadow-catcher.

File: src/components/Lights/Lights.component.tsx [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+
+  const [targets] = useState<{ head: Object3D; shoe: Object3D }>({
+    head: new Object3D(),
+    shoe: new Object3D()
+  });
+
+  useEffect(() => {
+    // apply provided positions for targets
+    targets.head.position.copy(lightTarget);
+    targets.shoe.position.set(0.0, 0.0, 0.0);
```
Comment (Zaehiel) [inline src/components/Lights/Lights.component.tsx:58]: Nice idea!

**Review by Zaehiel — CHANGES_REQUESTED** [review]

File: src/components/Lights/Lights.component.tsx (hunk 11 above)
Comment (ericjansenrpm) [reply src/components/Lights/Lights.component.tsx:99]: Maybe leaving the old shadow is more performant than the contact shadows. Especially having already optional AO in place which can achieve a similar look.

File: src/components/Lights/Lights.component.tsx (hunk 10 above)
Comment (ericjansenrpm) [reply src/components/Lights/Lights.component.tsx:46]: Yes I agree that this is way easier than the additional validation with defined props. Since the Avatar component exposed all the merged props flat to the user it sometimes can happen that some are explicitly undefined. Before we passed in default props in the top level to avoid that. I prefer having the safety fallback right there in the component to prevent misuse.

File: src/components/Lights/Lights.component.tsx (hunk 9 above)
Comment (ericjansenrpm) [reply src/components/Lights/Lights.component.tsx:32]: We use this polyfill for prop checking in the Lights component.

File: src/services/Models.service.tsx (hunk 8 above)
Comment (ericjansenrpm) [reply src/services/Models.service.tsx:85]: During mobtesting we noticed that the hair looked very plasticy with the new rim lights. We decided to go for an easy adjust here in order to avoid updating the assets. I think the defined value is something like .7 - .8 so it is only a minor tweak with decreasing the left light intensity.

File: src/components/Lights/Lights.component.tsx (hunk 12 above)
Comment (ericjansenrpm) [reply src/components/Lights/Lights.component.tsx:57]: 🙏

File: src/components/Lights/Lights.component.tsx (hunk 11 above)
Comment (ericjansenrpm) [reply src/components/Lights/Lights.component.tsx]: Removed contact shadows for now.

File: src/components/Lights/Lights.component.tsx (hunk 9 above)
Comment (Zaehiel) [reply src/components/Lights/Lights.component.tsx:31]: Let's do it in a way that doesn't override the default Javascript method `Object.fromEntries` 🙏

File: src/components/Avatar/Avatar.component.tsx [hunk 13]
```diff
@@ -272,23 +264,23 @@ const Avatar: FC<AvatarProps> = ({
       {background?.src && <Box {...background} />}
```
Comment (Zaehiel) [inline src/components/Avatar/Avatar.component.tsx:264]: Shadows no longer work, as the spotlights aren't set to cast shadows.
Also, turning on `castShadow` for one of the spotlights doesn't give the intended result.

Screenshot 2024-01-02 at 15 22 25

vs

Screenshot 2024-01-02 at 15 22 44

Comment (ericjansenrpm) [reply src/components/Avatar/Avatar.component.tsx:264]: The spotlights may not have the right rotation to reproduce the original shadow angle. A workaround could be to use a dummy directional light to cast the shadow in the desired direction without an intensity.

Screenshot from 2024-01-02 14-33-05

File: src/services/Models.service.tsx [hunk 14]
```diff
@@ -79,7 +79,10 @@ export const normaliseMaterialsConfig = (materials: Record<string, Material>, bl
     if (mat.map) {
       mat.map.minFilter = LinearFilter;
       mat.depthWrite = true;
-      mat.toneMapped = false;
+    }
```
[the remark this answers is not in the document]
Comment (ericjansenrpm) [reply src/services/Models.service.tsx:82]: Toggling this for emissive materials only is a great alternative. :) Thanks for the idea! The only small payoff is that those materials have an altered contrast compared to the tonemapped ones. But I think the difference is fine as long as we get the nice skintone improvement by having aces there doing its magic. :D

**Comment by Zaehiel** [comment]
@ericjansenrpm  I think we're missing a BREAKING CHANGES commit that would bump the major version. We're breaking the `<Avatar>` Props interface from previous versions

**Comment by ericjansenrpm** [comment]
Added a new commit with the BREAKING CHANGE content

**Comment by Zaehiel** [comment]
One last thing I want to check up on.
The current lighting setup is darker than the initial example we had.

I would say we'd need to put a bit more light back into the scene so hair is more it wouldn't look too dark.
It's very difficult to distinguish darker outfits with a dark background with current configuration.

Current:
Screenshot 2024-01-04 at 10 08 40


Expected:
Screenshot 2024-01-04 at 10 09 30


The silhouettes of the hands and feet are lost in comparison here:
Screenshot 2024-01-04 at 10 14 09

Screenshot 2024-01-04 at 10 14 16

**Comment by ericjansenrpm** [comment]
Screenshot from 2024-01-04 11-31-52
Screenshot from 2024-01-04 11-31-59
Screenshot from 2024-01-04 11-32-08

I have added a silhouette light and lifted the general brightness a bit by adjusting the tonemapping, further the colors are now aligned closer with the render api by using the hex colors from the blender scene.

**Review by Zaehiel — APPROVED** [review]
